File: 179-largest-number/largest-number.cpp
class Solution {
public:
    string largestNumber(vector<int>& nums) {
        vector<string> numbers;
        for (int num : nums)
            numbers.push_back(to_string(num));

        auto compare = [](const string& x, const string& y) {
            string xy = x + y;
            string yx = y + x;
            return xy > yx;
        };

        sort(numbers.begin(), numbers.end(), compare);
        string result;
        for (string t : numbers)
            result += t;
        return result[0] == '0' ? "0" : result;
    }
};

File: 179-largest-number/largest-number.java
import java.util.Arrays;
import java.util.Comparator;

class Solution {
    public String largestNumber(int[] nums) {
        String[] numbers = new String[nums.length];
        for (int i = 0; i < nums.length; i++)
            numbers[i] = String.valueOf(nums[i]);

        Comparator<String> compare = (x, y) -> {
            String xy = x + y;
            String yx = y + x;
            return yx.compareTo(xy);
        };

        Arrays.sort(numbers, compare);
        StringBuilder sb = new StringBuilder();
        for (String i : numbers)
            sb.append(i);
        return sb.charAt(0) == '0' ? "0" : sb.toString();
    }
}